crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_class.rs

```diff
@@ -0,0 +1,158 @@
+/// private type variables.
+///
+/// This rule only applies to generic classes and does not include generic functions. See
+/// [`non-pep695-generic-function`](non-pep695-generic-function.md) for the function version.
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+/// [PEP 696]: https://peps.python.org/pep-0696/
+#[derive(ViolationMetadata)]
+pub(crate) struct NonPEP695GenericClass {
+    name: String,
+}
+
+impl Violation for NonPEP695GenericClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Always;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NonPEP695GenericClass { name } = self;
+        format!("Generic class `{name}` uses `Generic` subclass instead of type parameters")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use type parameters".to_string())
+    }
+}
+
+/// UP046
+pub(crate) fn non_pep695_generic_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    // PEP-695 syntax is only available on Python 3.12+
+    if checker.settings.target_version < PythonVersion::Py312 {
+        return;
+    }
+
+    // don't try to handle generic classes inside other functions or classes
+    if in_nested_context(checker) {
+        return;
+    }
+
+    let StmtClassDef {
+        name,
+        type_params,
+        arguments,
+        ..
+    } = class_def;
+
+    // it's a runtime error to mix type_params and Generic, so bail out early if we see existing
+    // type_params
+    if type_params.is_some() {
+        return;
+    }
+
+    let Some(arguments) = arguments.as_ref() else {
+        return;
+    };
+
+    // TODO(brent) only accept a single, Generic argument for now. I think it should be fine to have
+    // other arguments, but this simplifies the fix just to delete the argument list for now
+    let [Expr::Subscript(ExprSubscript {
+        value,
+        slice,
+        range,
+        ..
+    })] = arguments.args.as_ref()
+    else {
+        return;
+    };
+
+    if !checker.semantic().match_typing_expr(value, "Generic") {
+        return;
+    }
+
+    let vars = {
+        let mut visitor = TypeVarReferenceVisitor {
+            vars: vec![],
+            semantic: checker.semantic(),
+        };
+        visitor.visit_expr(slice);
+        visitor.vars
+    };
+
+    let Some(type_vars) = check_type_vars(vars) else {
+        return;
+    };
+
+    // build the fix as a String to avoid removing comments from the entire function body
+    let type_params = DisplayTypeVars {
+        type_vars: &type_vars,
+        source: checker.source(),
+    };
+
+    checker.diagnostics.push(
+        Diagnostic::new(
+            NonPEP695GenericClass {
+                name: name.to_string(),
+            },
+            *range,
+        )
+        .with_fix(Fix::applicable_edit(
+            Edit::replacement(type_params.to_string(), name.end(), arguments.end()),
+            Applicability::Safe,
+        )),
```

Comment:
the fixes to these rules should be marked as unsafe rather than safe:
- The issue pointed out in the `Known issues` section regarding the switch from explicit variance to inferred variance means that they could change the semantics of the code's annotations
- the issue I pointed out just now regarding type variables from other modules means that the fixes could in some cases outright break the code's annotations

```suggestion
        .with_fix(Fix::unsafe_edit(Edit::replacement(
            type_params.to_string(),
            name.end(),
            arguments.end(),
        ))),
```

We also need to add a `## Fix safety` section to the rules' docs explaining why the fixes are marked as unsafe

Reply:
(if you grep for `## Fix safety`, you can find lots of other rules that have similar sections in their docs, that you can use as examples!)